chacra/async/rpm.py

```diff
@@ -73,16 +73,24 @@ def create_rpm_repo(repo_id):
         except OSError:
             logger.exception('could not symlink')
 
-    for d in repo_dirs:
-        # this prevents RPM packages that are larger than 2GB (!!!) from
-        # causing the database to fail to store the size and subsequently make
-        # the package uninstallable. Ideally, this types of flag options should
-        # be configurable
-        subprocess.check_call(['createrepo', '--no-database', d])
+    _do_createrepo(paths['absolute'], repo_dirs, repo.distro)
 
     logger.info("finished processing repository: %s", repo)
     repo.is_updating = False
     models.commit()
     timer.stop()
     counter += 1
     post_ready(repo)
+
+
+def _do_createrepo(base_path, repo_dirs, distro):
+    if distro in ['opensuse']:
```

Comment:
['opensuse', 'sle'] ?

Reply:
Done

Reply:
Good question. It got me wondering how chacra sets `distro` to "opensuse" in the opensuse case.

I found this (in `controllers/repos/__init__.py`):

```
from pecan import expose, abort, request
```

```
class FlavorsController(object):

    def __init__(self):
        self.distro_version = request.context['distro_version']
        self.project = Project.get(request.context['project_id'])
        self.distro_name = request.context['distro']
        self.ref = request.context['ref']
        self.sha1 = request.context['sha1']
        self.repos = self.project.repos.filter_by(
            distro=self.distro_name,
            distro_version=self.distro_version,
            ref=self.ref,
            sha1=self.sha1,
        ).all()
```

So it looks like this string is coming from `pecan` ?

Reply:
I guess `pecan` is populating the `distro` based on whatever is set in the HTTP request it is handling.

Reply:
yes.
Let's upload a binary:
```
(venv) vagrant@vagranthost:~$ chacractl binary create ceph/master/sha555/opensuse/15.1/noarch/flavors/default /var/tmp/build-root/openSUSE_Leap_15.1-x86_64/home/abuild/rpmbuild/RPMS/noarch/python3-oslo.cache-1.37.0-0.noarch.rpm
[urllib3.connectionpool][DEBUG ] Starting new HTTP connection (1): localhost:8000
[urllib3.connectionpool][DEBUG ] http://localhost:8000 "HEAD /binaries/ceph/master/sha555/opensuse/15.1/noarch/flavors/default/python3-oslo.cache-1.37.0-0.noarch.rpm/ HTTP/1.1" 404 0
[chacractl.api.binaries][INFO  ] POSTing file: /var/tmp/build-root/openSUSE_Leap_15.1-x86_64/home/abuild/rpmbuild/RPMS/noarch/python3-oslo.cache-1.37.0-0.noarch.rpm
[urllib3.connectionpool][DEBUG ] Starting new HTTP connection (1): localhost:8000
[urllib3.connectionpool][DEBUG ] http://localhost:8000 "POST /binaries/ceph/master/sha555/opensuse/15.1/noarch/flavors/default/ HTTP/1.1" 201 2
[urllib3.connectionpool][DEBUG ] Starting new HTTP connection (1): localhost:8000
[urllib3.connectionpool][DEBUG ] http://localhost:8000 "GET /binaries/ceph/master/sha555/opensuse/15.1/noarch/flavors/default/ HTTP/1.1" 200 656
```

Then check the database distro field:
```
postgres@vagranthost:~$ psql -d chacra -c "SELECT * FROM binaries WHERE distro='opensuse' and sha1='sha555';"|cat
 id |                  name                  |                                                     path                                                     |  ref   |  sha1  |  distro  | distro_version |  arch  | flavor  | built_by |          created           |          modified          | signed | size  |                                                             checksum                                                             | project_id | repo_id 
----+----------------------------------------+--------------------------------------------------------------------------------------------------------------+--------+--------+----------+----------------+--------+---------+----------+----------------------------+----------------------------+--------+-------+----------------------------------------------------------------------------------------------------------------------------------+------------+---------
  8 | python3-oslo.cache-1.37.0-0.noarch.rpm | /opt/binaries/ceph/master/sha555/opensuse/15.1/noarch/flavors/default/python3-oslo.cache-1.37.0-0.noarch.rpm | master | sha555 | opensuse | 15.1           | noarch | default |          | 2019-12-12 15:54:09.207212 | 2019-12-12 15:54:09.213632 | f      | 85588 | fd6a33fc4b78f0c6890c6aa6007a294e9124b5dfcefc510400946cfb071fb5c3b80d76dd75a73d8b80335d7c7ebc13cfa2a862250fb49ea3f97a3c71b9a5441b |          2 |       8
(1 row)
```

Reply:
@toabctl OK. Have you tried to upload a SLE binary?

Reply:
@smithfarm no. But you can upload any RPM to any distro. It's just the url which sets the `distro` field. Now using `tomsdistro`:
```
$ chacractl binary create ceph/master/sha555/tomsdistro/15.1/noarch/flavors/default /var/tmp/build-root/openSUSE_Leap_15.1-x86_64/home/abuild/rpmbuild/RPMS/noarch/python3-oslo.cache-1.37.0-0.noarch.rpm
```
and then in the DB:
```
postgres@vagranthost:~$ psql -d chacra -c "SELECT * FROM binaries WHERE distro='tomsdistro' and sha1='sha555';"|cat
 id |                  name                  |                                                      path                                                      |  ref   |  sha1  |   distro   | distro_version |  arch  | flavor  | built_by |          created           |          modified          | signed | size  |                                                             checksum                                                             | project_id | repo_id 
----+----------------------------------------+----------------------------------------------------------------------------------------------------------------+--------+--------+------------+----------------+--------+---------+----------+----------------------------+----------------------------+--------+-------+----------------------------------------------------------------------------------------------------------------------------------+------------+---------
  9 | python3-oslo.cache-1.37.0-0.noarch.rpm | /opt/binaries/ceph/master/sha555/tomsdistro/15.1/noarch/flavors/default/python3-oslo.cache-1.37.0-0.noarch.rpm | master | sha555 | tomsdistro | 15.1           | noarch | default |          | 2019-12-12 16:01:47.059632 | 2019-12-12 16:01:47.065884 | f      | 85588 | fd6a33fc4b78f0c6890c6aa6007a294e9124b5dfcefc510400946cfb071fb5c3b80d76dd75a73d8b80335d7c7ebc13cfa2a862250fb49ea3f97a3c71b9a5441b |          2 |       9
(1 row)

```

Reply:
And the distro for opensuse is set in https://github.com/ceph/ceph-build/blob/master/ceph-dev-build/build/setup_osc#L81 and then used in build_osc

Reply:
@smithfarm I don't see any practical case of using sle upstream at the moment, but this will be good to have when we (mainly me) want to setup our custom chacra servers :-)